src/cmd/connect.go

```diff
@@ -16,100 +16,121 @@ import (
-		defer spinner.Stop()
+}
 
-		c, err := cluster.NewCluster()
-		if err != nil {
-			return err
-		}
+// NewConnectCommand creates the `connect` sub-command and its nested children.
+func NewConnectCommand() *cobra.Command {
+	o := &ConnectOptions{}
 
-		var tunnel *cluster.Tunnel
-		if target == "" {
-			tunnel, err = c.ConnectTunnelInfo(ctx, zt)
-		} else {
-			var ti cluster.TunnelInfo
-			ti, err = c.NewTargetTunnelInfo(ctx, target)
-			if err != nil {
-				return fmt.Errorf("unable to create tunnel: %w", err)
-			}
-			if zt.LocalPort != 0 {
-				ti.LocalPort = zt.LocalPort
-			}
-			tunnel, err = c.ConnectTunnelInfo(ctx, ti)
-		}
+	cmd := &cobra.Command{
+		Use:     "connect { REGISTRY | GIT | connect-name }",
+		Aliases: []string{"c"},
+		Short:   lang.CmdConnectShort,
+		Long:    lang.CmdConnectLong,
+		RunE:    o.Run,
+	}
 
-		if err != nil {
-			return fmt.Errorf("unable to connect to the service: %w", err)
-		}
+	cmd.Flags().StringVar(&o.zt.ResourceName, "name", "", lang.CmdConnectFlagName)
+	cmd.Flags().StringVar(&o.zt.Namespace, "namespace", cluster.ZarfNamespaceName, lang.CmdConnectFlagNamespace)
+	cmd.Flags().StringVar(&o.zt.ResourceType, "type", cluster.SvcResource, lang.CmdConnectFlagType)
+	cmd.Flags().IntVar(&o.zt.LocalPort, "local-port", 0, lang.CmdConnectFlagLocalPort)
+	cmd.Flags().IntVar(&o.zt.RemotePort, "remote-port", 0, lang.CmdConnectFlagRemotePort)
+	cmd.Flags().BoolVar(&o.cliOnly, "cli-only", false, lang.CmdConnectFlagCliOnly)
 
-		defer tunnel.Close()
-
-		if cliOnly {
-			spinner.Updatef(lang.CmdConnectEstablishedCLI, tunnel.FullURL())
-			l.Info("Tunnel established, waiting for user to interrupt (ctrl-c to end)", "url", tunnel.FullURL())
-		} else {
-			spinner.Updatef(lang.CmdConnectEstablishedWeb, tunnel.FullURL())
-			l.Info("Tunnel established, opening your default web browser (ctrl-c to end)", "url", tunnel.FullURL())
-			if err := exec.LaunchURL(tunnel.FullURL()); err != nil {
-				return err
-			}
-		}
+	// TODO(soltysh): consider splitting sub-commands into separate files
+	cmd.AddCommand(NewConnectListCommand())
 
-		// Wait for the interrupt signal or an error.
-		select {
-		case <-ctx.Done():
-			spinner.Successf(lang.CmdConnectTunnelClosed, tunnel.FullURL())
-			return nil
-		case err = <-tunnel.ErrChan():
-			return fmt.Errorf("lost connection to the service: %w", err)
-		}
-	},
+	return cmd
 }
 
-var connectListCmd = &cobra.Command{
-	Use:     "list",
-	Aliases: []string{"l"},
-	Short:   lang.CmdConnectListShort,
-	RunE: func(cmd *cobra.Command, _ []string) error {
-		c, err := cluster.NewCluster()
+// Run performs the execution of 'connect' sub command.
+func (o *ConnectOptions) Run(cmd *cobra.Command, args []string) error {
+	ctx := cmd.Context()
+	l := logger.From(ctx)
+	target := ""
+	if len(args) > 0 {
+		target = args[0]
+	}
+
+	spinner := message.NewProgressSpinner(lang.CmdConnectPreparingTunnel, target)
+	defer spinner.Stop()
+
+	c, err := cluster.NewCluster()
```

Comment:
Definitely not in scope for this PR, but sometime in the future it'd be great to ensure every CLI command can call a clearly defined API fn that encapsulates all of the configuration. Connect in particular has a lot of logic that could be abstracted out of the cli layer, which made me think of this.